compiler/desugared/from_surface.ml

```diff
@@ -553,177 +538,127 @@ let rec translate_expr
     in
     Expr.eapp
       (Expr.eop
-         (match op' with
-         | Surface.Ast.Map -> Map
-         | Surface.Ast.Filter -> Filter
-         | _ -> assert false (* should not happen *))
+         (match op with
+         | S.Map _ -> Map
+         | S.Filter _ -> Filter
+         | _ -> assert false)
          [TAny, pos; TAny, pos]
          emark)
       [f_pred; collection] emark
   | CollectionOp
-      ( ( Surface.Ast.Aggregate
-            (Surface.Ast.AggregateArgExtremum (max_or_min, pred_typ, init)),
-          pos_op' ),
-        param',
-        collection,
-        predicate ) ->
-    let init = rec_helper init in
+      (S.AggregateArgExtremum { max; default; f = param, predicate }, collection)
+    ->
+    let default = rec_helper default in
+    let pos_dft = Expr.pos default in
     let collection = rec_helper collection in
     let ctxt, param =
-      Name_resolution.add_def_local_var ctxt (Marked.unmark param')
+      Name_resolution.add_def_local_var ctxt (Marked.unmark param)
     in
-    let op_ty = aggregate_typ pos pred_typ in
-    let cmp_op = if max_or_min then Op.Gt else Op.Lt in
+    let cmp_op = if max then Op.Gt else Op.Lt in
     let f_pred =
       Expr.make_abs [| param |]
         (translate_expr scope inside_definition_of ctxt predicate)
         [TAny, pos]
         pos
     in
-    let acc_var = Var.make "acc" in
-    let acc_var_e = Expr.make_var acc_var emark in
-    let item_var = Var.make "item" in
-    let item_var_e = Expr.make_var item_var (Marked.get_mark collection) in
-    let fold_body =
-      Expr.eifthenelse
-        (Expr.eapp
-           (Expr.eop cmp_op
-              [op_ty, pos_op'; op_ty, pos_op']
-              (Untyped { pos = pos_op' }))
-           [
-             Expr.eapp f_pred [acc_var_e] emark;
-             Expr.eapp f_pred [item_var_e] emark;
-           ]
-           emark)
-        acc_var_e item_var_e emark
-    in
-    let fold_f =
-      Expr.make_abs [| acc_var; item_var |] fold_body [TAny, pos; TAny, pos] pos
+    let v1, v2 = Var.make "x1", Var.make "x2" in
+    let x1 = Expr.make_var v1 emark in
+    let x2 = Expr.make_var v2 emark in
+    let reduce_f =
+      (* Note: this computes f_pred twice on every element, but we'd rather not
```

Comment:
Indeed that would be better, laziness I guess, I kept it as it was. Note however that the better solution would require tuples that aren't available (atm) at that stage of compilation, so that wouldn't be trivial.

Since they exist in dcalc, they just be to be generalised though.